runtime/shared_common/CacheLifecycleManager.cpp

```diff
@@ -782,24 +797,33 @@ j9shr_destroy_cache(struct J9JavaVM* vm, const char* ctrlDirName, UDATA verboseF
 	}
 	/* try to remove older generation cache first */
 	for (UDATA i = generationStart; i <= lastGeneration; i++) {
-		/* explicitly pass 0 as verboseFlags to j9shr_stat_cache to suppress printing of messages related to cache existence in all cases */
-		if (1 == j9shr_stat_cache(vm, cacheDirName, 0 , cacheName, versionData, i)) {
-			cacheStatus = J9SH_DESTROYED_OLDER_GEN_CACHE;
-			SH_OSCache::newInstance(PORTLIB, oscache, cacheName, i, versionData);
-			/*Pass in 0 for runtime flags means we will not check semid in cache header when we destroy a cache*/
-			if (!oscache->startup(vm, ctrlDirName, vm->sharedCacheAPI->cacheDirPerm, cacheName, vm->sharedClassPreinitConfig, 0, J9SH_OSCACHE_OPEXIST_DESTROY, verboseFlags, 0/*runtime check*/, 0, vm->sharedCacheAPI->storageKeyTesting, versionData, NULL, SHR_STARTUP_REASON_DESTROY)) {
-				if (oscache->getError() != J9SH_OSCACHE_NO_CACHE) {
-					cacheStatus = J9SH_DESTROY_FAILED_OLDER_GEN_CACHE;
+		I_8 layerMaxOldGen = -1;
+		I_8 layerMinOldGen = -1;
+		if (i > J9SH_GENERATION_37) {
+			/* J9SH_DESTROY_TOP_LAYER_ONLY take effects on current generation only. Remove all layers of older generation. */
+			layerMaxOldGen = (J9SH_DESTROY_TOP_LAYER_ONLY == layerEnd) ? J9SH_LAYER_NUM_MAX_VALUE : layerEnd;
```

Comment:
I'm thinking old generations should just delete all layers.

Reply:
Right, it is fixed.